### ISSUE TITLE
alien desease: psychiatrist send home the patient.

### ISSUE THREAD

**Issue by overlordchaos** [CorsixTH/CorsixTH#1303]
#### Describe the issue
every time a psychiatrist takes the place of a GP's office, when an alien visits them, he sends them home and the reason is "there is no more avaidable diagnosis rooms" considering i have a hospital with ALL INGAME ONES its impossible, not to mention the DNA machine will cure them. also i have noticed regular doctor's/surgeon's to send the patient STRAIGHT to the DNA machine to get cured. so regular doctor's will not get cured just send directly into the threatment room without any diagnosis whatsoever.
#### Steps to Reproduce
1. start any level with the alien desease and all the diagnosis/consultation rooms avaidable.

2.wait for an alien patient to arrive (if it is not enabled to have random alien patients then wait untill a emergency is triggered?)

3.watch when any psychiatrist its on the GP office. they will kick them out of the hospital.

#### Expected Behavior
(as any other doctor/surgeon) make him to go to get diagnosis them cure them.
#### System Information
**CorsixTH Version:** *e.g. 0.50 or 7d886f35ca*
0.61 RC 1
**Operating System:** *e.g. Windows 10*
win 10 x64
**Theme Hospital Version:** *CD, GOG.com, Origin, or Demo*
gog.com
#### Gamelog.txt
Created game with savegame version 121.
Error in timer handler: 
C:\Program Files\CorsixTH\Lua\entities\patient.lua:182: attempt to perform arithmetic on a string value (field 'MaxDiagDiff')
stack traceback:
	C:\Program Files\CorsixTH\Lua\entities\patient.lua:182: in method 'completeDiagnosticStep'
	C:\Program Files\CorsixTH\Lua\rooms\gp.lua:135: in method 'dealtWithPatient'
	C:\Program Files\CorsixTH\Lua\rooms\gp.lua:81: in method 'loop_callback'
	...ogram Files\CorsixTH\Lua\humanoid_actions\use_object.lua:324: in local 'timer_function'
	C:\Program Files\CorsixTH\Lua\entity.lua:216: in field 'tick'
	C:\Program Files\CorsixTH\Lua\entities\staff.lua:149: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1051: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1008: in function <C:\Program Files\CorsixTH\Lua\app.lua:1005>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:909: in function <C:\Program Files\CorsixTH\Lua\app.lua:904>
Recovering from error in timer handler...

Empty action queue!
Last action: walk
stack traceback:
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:544: in upvalue 'Humanoid_startAction'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:670: in method 'finishAction'
	C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:193: in function <C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:178>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\entity.lua:216: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1051: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1008: in function <C:\Program Files\CorsixTH\Lua\app.lua:1005>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:909: in function <C:\Program Files\CorsixTH\Lua\app.lua:904>
The humanoid was told to leave the hospital...
Savegame version is 121 (v0.61-rc1), originally it was 121 (v0.61-rc1)
Error in timer handler: 
C:\Program Files\CorsixTH\Lua\entities\patient.lua:950: bad argument #1 to 'getCellFlags' (Map co-ordinates out of bounds)
stack traceback:
	[C]: in method 'getCellFlags'
	C:\Program Files\CorsixTH\Lua\entities\patient.lua:950: in method 'setTile'
	C:\Program Files\CorsixTH\Lua\entity.lua:171: in method 'setTilePositionSpeed'
	C:\Program Files\CorsixTH\Lua\humanoid_actions\spawn.lua:115: in field '?'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:578: in upvalue 'Humanoid_startAction'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:645: in method 'setNextAction'
	C:\Program Files\CorsixTH\Lua\objects\helicopter.lua:90: in method 'spawnPatient'
	C:\Program Files\CorsixTH\Lua\objects\helicopter.lua:66: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1051: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1008: in function <C:\Program Files\CorsixTH\Lua\app.lua:1005>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:909: in function <C:\Program Files\CorsixTH\Lua\app.lua:904>
Recovering from error in timer handler...

Empty action queue!
Last action: spawn
stack traceback:
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:544: in upvalue 'Humanoid_startAction'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:670: in local 'timer_function'
	C:\Program Files\CorsixTH\Lua\entity.lua:216: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1051: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1008: in function <C:\Program Files\CorsixTH\Lua\app.lua:1005>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:909: in function <C:\Program Files\CorsixTH\Lua\app.lua:904>
Error in timer handler: 
...m Files\CorsixTH\Lua\humanoid_actions\seek_reception.lua:42: humanoid must be associated with a hospital to seek reception
stack traceback:
	[C]: in function 'assert'
	...m Files\CorsixTH\Lua\humanoid_actions\seek_reception.lua:42: in field '?'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:578: in upvalue 'Humanoid_startAction'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:670: in local 'timer_function'
	C:\Program Files\CorsixTH\Lua\entity.lua:216: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1051: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1008: in function <C:\Program Files\CorsixTH\Lua\app.lua:1005>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:909: in function <C:\Program Files\CorsixTH\Lua\app.lua:904>
Recovering from error in timer handler...
Error in timer handler: 
C:\Program Files\CorsixTH\Lua\entities\patient.lua:182: attempt to perform arithmetic on a string value (field 'MaxDiagDiff')
stack traceback:
	C:\Program Files\CorsixTH\Lua\entities\patient.lua:182: in method 'completeDiagnosticStep'
	C:\Program Files\CorsixTH\Lua\rooms\gp.lua:135: in method 'dealtWithPatient'
	C:\Program Files\CorsixTH\Lua\rooms\gp.lua:81: in method 'loop_callback'
	...ogram Files\CorsixTH\Lua\humanoid_actions\use_object.lua:324: in local 'timer_function'
	C:\Program Files\CorsixTH\Lua\entity.lua:216: in field 'tick'
	C:\Program Files\CorsixTH\Lua\entities\staff.lua:149: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1051: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1008: in function <C:\Program Files\CorsixTH\Lua\app.lua:1005>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:909: in function <C:\Program Files\CorsixTH\Lua\app.lua:904>
Recovering from error in timer handler...

Empty action queue!
Last action: walk
stack traceback:
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:544: in upvalue 'Humanoid_startAction'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:670: in method 'finishAction'
	C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:193: in function <C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:178>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\entity.lua:216: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1051: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1008: in function <C:\Program Files\CorsixTH\Lua\app.lua:1005>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:909: in function <C:\Program Files\CorsixTH\Lua\app.lua:904>
The humanoid was told to leave the hospital...
Error in timer handler: 
C:\Program Files\CorsixTH\Lua\entities\patient.lua:182: attempt to perform arithmetic on a string value (field 'MaxDiagDiff')
stack traceback:
	C:\Program Files\CorsixTH\Lua\entities\patient.lua:182: in method 'completeDiagnosticStep'
	C:\Program Files\CorsixTH\Lua\rooms\gp.lua:135: in method 'dealtWithPatient'
	C:\Program Files\CorsixTH\Lua\rooms\gp.lua:81: in method 'loop_callback'
	...ogram Files\CorsixTH\Lua\humanoid_actions\use_object.lua:324: in local 'timer_function'
	C:\Program Files\CorsixTH\Lua\entity.lua:216: in field 'tick'
	C:\Program Files\CorsixTH\Lua\entities\staff.lua:149: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1051: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1008: in function <C:\Program Files\CorsixTH\Lua\app.lua:1005>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:909: in function <C:\Program Files\CorsixTH\Lua\app.lua:904>

### Additional Info
maybe this is an original game perk from the devs (in real life if a patient says "hey doctor i have DNA alien on my body and its making me sick" would go straight to the nut house) but this is ridicolous. i cannot have psychiatrists on the GP's offices at any time cause i cannot trust them with DNA alien patients.
EDIT: on the RC version, ALL the doctor's on the GP office will send away to the alien patients with the excuse that there are no more avaidable exam rooms...

**Comment by wolfy1339** [comment]
Does this apply to psychiatrists that are also surgeons or researchers or both?

**Comment by overlordchaos** [comment]
@wolfy1339 up to my understanding, if it is a psychiatrist wont matter if they are surgeon's/researcher's as long as if it a psychiatrist, with or without other specialty will kick the patient out cause "there are no more diag rooms avaidable" at least thats the reason if you put the cursor over the kicked alien every time.

**Comment by TheCycoONE** [comment]
Start a new game with the corrected `MaxDiagDiff` and see if you can still reproduce. It seems likely to be connected to that (and I can see from your game log that there's still invalid values there). Note that changing a level and opening the old save will not work.

**Comment by overlordchaos** [comment]
@TheCycoONE changed some values in the range of 100 to 1000 for maxdiagdiff. i will copy and paste to see the updated values:
#expertise[1].Known.RschReqd 1 0 GENERAL_PRACTICE
#expertise[2].Known.RschReqd.MaxDiagDiff 1 0 150 BLOATY_HEAD
#expertise[3].Known.RschReqd.MaxDiagDiff 1 0 350 HAIRYITUS
#expertise[4].Known.RschReqd.MaxDiagDiff 1 0 100 ELVIS
#expertise[5].Known.RschReqd.MaxDiagDiff 1 0 100 INVIS
#expertise[6].Known.RschReqd.MaxDiagDiff 1 0 900 RADIATION
#expertise[7].Known.RschReqd.MaxDiagDiff 1 0 150 SLACK_TONGUE
#expertise[8].Known.RschReqd.MaxDiagDiff 1 0 900 ALIEN
#expertise[9].Known.RschReqd.MaxDiagDiff 1 0 350 BROKEN_BONES
#expertise[10].Known.RschReqd.MaxDiagDiff 1 0 750 BALDNESS
#expertise[11].Known.RschReqd.MaxDiagDiff 1 0 650 DISCRETE_ITCHING
#expertise[12].Known.RschReqd.MaxDiagDiff 1 0 800 JELLYITUS
#expertise[13].Known.RschReqd.MaxDiagDiff 1 0 500 SLEEPING_ILLNESS
#expertise[14].Known.RschReqd.MaxDiagDiff 1 0 150 PREGNANT
#expertise[15].Known.RschReqd.MaxDiagDiff 1 0 100 TRANSPARENCY
#expertise[16].Known.RschReqd.MaxDiagDiff 1 0 250 UNCOMMON_COLD
#expertise[17].Known.RschReqd.MaxDiagDiff 1 0 300 BROKEN_WIND
#expertise[18].Known.RschReqd.MaxDiagDiff 1 0 950 SPARE_RIBS
#expertise[19].Known.RschReqd.MaxDiagDiff 1 0 750 KIDNEY_BEANS
#expertise[20].Known.RschReqd.MaxDiagDiff 1 0 700 BROKEN_HEART
#expertise[21].Known.RschReqd.MaxDiagDiff 1 0 850 RUPTURED_NODULES
#expertise[22].Known.RschReqd.MaxDiagDiff 1 0 990 MULTIPLE_TV_PERSONALITIES
#expertise[23].Known.RschReqd.MaxDiagDiff 1 0 300 INFECTIOUS_LAUGHTER
#expertise[24].Known.RschReqd.MaxDiagDiff 1 0 200 CORRUGATED_ANKLES
#expertise[25].Known.RschReqd.MaxDiagDiff 1 0 400 CHRONIC_NOSEHAIR
#expertise[26].Known.RschReqd.MaxDiagDiff 1 0 600 3RD_DEGREE_SIDEBURNS
#expertise[27].Known.RschReqd.MaxDiagDiff 1 0 750 FAKE_BLOOD
#expertise[28].Known.RschReqd.MaxDiagDiff 1 0 300 GASTRIC_EJECTIONS
#expertise[29].Known.RschReqd.MaxDiagDiff 1 0 200 THE_SQUITS
#expertise[30].Known.RschReqd.MaxDiagDiff 1 0 950 IRON_LUNGS
#expertise[31].Known.RschReqd.MaxDiagDiff 1 0 200 SWEATY_PALMS
#expertise[32].Known.RschReqd.MaxDiagDiff 1 0 300 HEAPED_PILES
#expertise[33].Known.RschReqd.MaxDiagDiff 1 0 700 GUT_ROT
#expertise[34].Known.RschReqd.MaxDiagDiff 1 0 800 GOLF_STONES
#expertise[35].Known.RschReqd.MaxDiagDiff 1 0 950 UNEXPECTED_SWELLING
with this values ALL the aliens are kicked out. yet on a default level they are still kicked out. i dont think its exclusively to this values since on the beta 0.61 9 out of 10 patients are kicked out. on my RC test up to this point 10 out of 10 patients are kicked out. and even with incorrect values before and after the update.

**Comment by MarkL1961** [comment]
From memory you need a consultant to diagnose the first one, once it is a known disease it gets easier.
Oddly you can build the treatment before something is known, but that is the same with other diseases at times

Same in TH as well

**Comment by overlordchaos** [comment]
@MarkL1961 yes it was on 100 and  @TheCycoONE said it must be higher. 
my custom level enables ALL deseases/exam rooms/consults enabled by default since the start.
even if i build the research room and put a researcher, only random upgrades are developed. others are not, like the enhaced machines to add into the research room. maybe the desease appears on the casebook BUT the AI cannot actually use it?

**Comment by MarkL1961** [comment]
Take a look at the level file for level 1 in campaign, I have a feeling you have to be lower not higher.

Research is hard coded and not in the level file from what I remember,  what is in the level file is the ability to make everything available from the start

Have you read https://github.com/CorsixTH/CorsixTH/wiki/Custom-Levels?

**Comment by MarkL1961** [comment]
Unless things have changed here are some examples from a level file

-------------------- Disease Configuration -------------------------

When a drug is researched, what effectiveness does it have
#gbv.StartRating 80 

The following table contains all diagnoses and treatments that shows up in the drug casebook
in the game. Known specifies whether it should show up from the beginning of the level and
RschReqd how much research is required to discover the treatment room for the disease.

#expertise[1].Known.RschReqd 1 0 GENERAL_PRACTICE
#expertise[2].Known.RschReqd 1 40000 BLOATY_HEAD
#expertise[3].Known.RschReqd 0 40000 HAIRYITUS
#expertise[4].Known.RschReqd 0 60000 ELVIS
#expertise[5].Known.RschReqd 0 60000 INVIS
#expertise[6].Known.RschReqd 0 60000 RADIATION
#expertise[7].Known.RschReqd 0 40000 SLACK_TONGUE
#expertise[8].Known.RschReqd 0 60000 ALIEN
#expertise[9].Known.RschReqd 0 20000 BROKEN_BONES
#expertise[10].Known.RschReqd 0 40000 BALDNESS
#expertise[11].Known.RschReqd 0 40000 DISCRETE_ITCHING
#expertise[12].Known.RschReqd 0 40000 JELLYITUS
#expertise[13].Known.RschReqd 0 40000 SLEEPING_ILLNESS
#expertise[14].Known.RschReqd 0 30000 PREGNANT
#expertise[15].Known.RschReqd 0 40000 TRANSPARENCY
#expertise[16].Known.RschReqd 0 20000 UNCOMMON_COLD
#expertise[17].Known.RschReqd 0 60000 BROKEN_WIND
#expertise[18].Known.RschReqd 0 20000 SPARE_RIBS
#expertise[19].Known.RschReqd 0 20000 KIDNEY_BEANS
#expertise[20].Known.RschReqd 0 20000 BROKEN_HEART
#expertise[21].Known.RschReqd 0 20000 RUPTURED_NODULES
#expertise[22].Known.RschReqd 0 40000 MULTIPLE_TV_PERSONALITIES
#expertise[23].Known.RschReqd 1 60000 INFECTIOUS_LAUGHTER
#expertise[24].Known.RschReqd 0 40000 CORRUGATED_ANKLES
#expertise[25].Known.RschReqd 0 40000 CHRONIC_NOSEHAIR
#expertise[26].Known.RschReqd 0 40000 3RD_DEGREE_SIDEBURNS
#expertise[27].Known.RschReqd 0 40000 FAKE_BLOOD
#expertise[28].Known.RschReqd 0 40000 GASTRIC_EJECTIONS
#expertise[29].Known.RschReqd 1 20000 THE_SQUITS
#expertise[30].Known.RschReqd 0 20000 IRON_LUNGS
#expertise[31].Known.RschReqd 0 40000 SWEATY_PALMS
#expertise[32].Known.RschReqd 0 20000 HEAPED_PILES
#expertise[33].Known.RschReqd 0 20000 GUT_ROT
#expertise[34].Known.RschReqd 0 20000 GOLF_STONES
#expertise[35].Known.RschReqd 0 20000 UNEXPECTED_SWELLING
#expertise[36].Known.RschReqd 0 40000 I_D_SCANNER             DIAGNOSIS
#expertise[37].Known.RschReqd 0 50000 I_D_BLOOD_MACHINE       DIAGNOSIS
#expertise[38].Known.RschReqd 0 20000 I_D_CARDIO              DIAGNOSIS
#expertise[39].Known.RschReqd 0 30000 I_D_XRAY                DIAGNOSIS
#expertise[40].Known.RschReqd 0 60000 I_D_ULTRASCAN           DIAGNOSIS
#expertise[41].Known.RschReqd 1 20000 I_D_STANDARD            DIAGNOSIS
#expertise[42].Known.RschReqd 1 20000 I_D_WARD                DIAGNOSIS
#expertise[43].Known.RschReqd 1 20000 I_D_SHRINK              DIAGNOSIS

| Objects available | Available from the start | Strength | Available for this level | Comment
#objects[9].StartAvail.StartStrength.AvailableForLevel 1 8 1 Inflator Machine
#objects[13].StartAvail.StartStrength.AvailableForLevel 1 13 1 Cardiogram
#objects[14].StartAvail.StartStrength.AvailableForLevel 0 12 1 Scanner
#objects[22].StartAvail.StartStrength.AvailableForLevel 0 9 1 Ultrascan
#objects[23].StartAvail.StartStrength.AvailableForLevel 0 7 1 DNA Restorer
#objects[24].StartAvail.StartStrength.AvailableForLevel 0 11 1 Cast Remover
#objects[25].StartAvail.StartStrength.AvailableForLevel 0 8 1 Hair restorer
#objects[26].StartAvail.StartStrength.AvailableForLevel 0 10 1 Slicer for slack tongues
#objects[27].StartAvail.StartStrength.AvailableForLevel 0 12 1 X-Ray
#objects[30].StartAvail.StartStrength.AvailableForLevel 1 12 1 Operating Table
#objects[37].StartAvail.StartStrength.AvailableForLevel 1 0 1 Projector
#objects[40].StartAvail.StartStrength.AvailableForLevel	0 1 1 Research Computer
#objects[41].StartAvail.StartStrength.AvailableForLevel	0 1 1 Chemical Mixer 
#objects[42].StartAvail.StartStrength.AvailableForLevel 0 10 1 Blood Machine
#objects[46].StartAvail.StartStrength.AvailableForLevel 0 8 1 Electrolysis Machine
#objects[47].StartAvail.StartStrength.AvailableForLevel 0 7 1 Jellyitus Moulding Machine
#objects[54].StartAvail.StartStrength.AvailableForLevel 0 10 1 Decontamination Shower
#objects[55].StartAvail.StartStrength.AvailableForLevel 1 10 1 Autopsy Research Machine
#objects[57].StartAvail.StartStrength.AvailableForLevel 0 57 1 Video Game

the number of months before first
#visuals[0]  1 I_BLOATY_HEAD
#visuals[1]  21 I_HAIRYITUS
#visuals[2]  7 I_ELVIS
#visuals[3]  9 I_INVIS
#visuals[4]  23 I_RADIATION
#visuals[5]  11 I_SLACK_TONGUE
#visuals[6]  41 I_ALIEN
#visuals[7]  11 I_BROKEN_BONES
#visuals[8]  12 I_BALDNESS
#visuals[9]  5 I_DISCRETE_ITCHING
#visuals[10] 12 I_JELLYITUS
#visuals[11] 1 I_SLEEPING_ILLNESS
#visuals[12] 18 I_PREGNANT
#visuals[13] 16 I_TRANSPARENCY
#non_visuals[0] 1 I_UNCOMMON_COLD
#non_visuals[1] 3 I_BROKEN_WIND
#non_visuals[2] 19 I_SPARE_RIBS
#non_visuals[3] 31 I_KIDNEY_BEANS
#non_visuals[4] 25 I_BROKEN_HEART
#non_visuals[5] 18 I_RUPTURED_NODULES
#non_visuals[6] 1 I_MULTIPLE_TV_PERSONALITIES
#non_visuals[7] 14 I_INFECTIOUS_LAUGHTER
#non_visuals[8] 21 I_CORRUGATED_ANKLES
#non_visuals[9] 15 I_CHRONIC_NOSEHAIR
#non_visuals[10] 1 I_3RD_DEGREE_SIDEBURNS
#non_visuals[11] 31 I_FAKE_BLOOD
#non_visuals[12] 13 I_GASTRIC_EJECTIONS
#non_visuals[13] 6 I_THE_SQUITS
#non_visuals[14] 9 I_IRON_LUNGS
#non_visuals[15] 9 I_SWEATY_PALMS
#non_visuals[16] 15 I_HEAPED_PILES
#non_visuals[17] 18 I_GUT_ROT
#non_visuals[18] 7 I_GOLF_STONES
#non_visuals[19] 5 I_UNEXPECTED_SWELLING

**Comment by TheCycoONE** [comment]
@MarkL1961 what are you talking about? There's no `MaxDiffDiag` in your list, and that number is between 0 and 1000. I have never seen it set below 100.

**Comment by overlordchaos** [comment]
everyone can try to replicate this issue on 0.61 RC if you think this is a issue only from my custom level.
aldo i dont think so. it could be the desease was hard coded disabled cause they released the original game and has not been 100% fixed?

**Comment by mugmuggy** [comment]
Without looking at any saves and ignoring the MaxDiagDiff and looking only at the 'kicked' issue, I think normal behaviour is as follows;

By default alien patients are only enabled for emergency.
`  alien_dna_only_by_emergency = true,`

The alien dna disease has
`disease.only_emergency = TheApp.config.alien_dna_only_by_emergency -- TODO implement (there are no normal door animations, so they cannot go to GP)`

When a patient is spawned it checks to see if a patient can have that disease.
`local function isDiseaseUsableForNewPatient(self, disease, hospital)`
`  if disease.only_emergency then return false end`

If its an emergency patient, the patient is diagnosed already and can go to treatment room directly not ever visiting the GP.

Going back to the initial post,

> 2.wait for an alien patient to arrive (if it is not enabled to have random alien patients then wait untill a emergency is triggered?)

That would would mean `alien_dna_only_by_emergency = false`.

I assume then, the patient then goes into the dealtWithPatient process ending up in this if statement
https://github.com/CorsixTH/CorsixTH/blob/63afae1554a2abeba225d5865f5ea77b18e73c78/CorsixTH/Lua/rooms/gp.lua#L130-L150

Skipping to the else at the end, as diagnosis is in progress which does this.
https://github.com/CorsixTH/CorsixTH/blob/63afae1554a2abeba225d5865f5ea77b18e73c78/CorsixTH/Lua/rooms/gp.lua#L167-L172

As alien dna doesn't have any diagnosis rooms, this is taken and the patient is kicked.

**Comment by overlordchaos** [comment]
@mugmuggy so you are in part confirming what i report. yes on the beta i assume the GP office takes the part of the code that threats as a "emergency" patient and sends it straight to the DNA machine to be cured.
on the RC release its just kicked out ALL the visiting alien patients (not the emergency ones those are fine)
i have enabled visitor's alien's and the missing animations are made with regular men/woman spirtes then it switchs back to the alien form (to walk mostly) i am guessing it is nessesary to implement for visiting patients diagnosis rooms and then to be send to the DNA machine afterwards.
i would suggest the blood test room, then the general diag room then the DNA machine to be cured.

**Comment by TheCycoONE** [comment]
https://github.com/CorsixTH/CorsixTH/compare/v0.61-beta1...v0.61-rc1

Nothing relevant that I see.

Emergency patients don't walk in, they come by helicopter when you agree to handle an emergency.

**Comment by TheCycoONE** [comment]
Note that there is always a chance that the go can completely diagnose a patient. My feeling is in your beta save you just got lucky with a highly skilled GP.

That said you likely want to disable walk in alien DNA in your level.

**Comment by MarkL1961** [comment]
@TheCycoONE I never said it was in my list.  I said if you miss it off the game uses the defaults

Take a look at easy level 1 , unless my file is corrupted it is set as 1
#expertise[5].Known.RschReqd.MaxDiagDiff	1	0	1	INVIS

**Comment by MarkL1961** [comment]
Alien DNA do not attend other diagnostic rooms that is how it was in TH 
It is disabled in the game by default as there are no animations for siting or using a door etc.  If you enable it you accept that they will change into men/women and then change back.  This was how it was in TH.
You need a consultant or a GP that is nearly a consultant/has high skills for it to be diagnosed.
Note that there are also issues with the animations for the treatment

It has always been this way

**Comment by TheCycoONE** [comment]
@MarkL1961 sorry I misunderstood. Yes you're right, the early levels do seem to have some very low values, e.g. the 1 for INVIS (at all difficulties), and 10 for INVIS in level 3 etc.

I have a very strong feeling that CorsixTH is not handling that value the same way that the original game did. A value of 1 should (I think) mean that the GP can always diagnose the disease regardless of skill - instead we are using it as a modifer on the base diagnostic progress (capped at 0.4) which really diminishes the impact of MaxDiagDiff, so that it makes at most at 0.4 difference in diagnostic progress:
https://github.com/CorsixTH/CorsixTH/blob/4349cc74d2fb07dbca02f4c1f0678d1b7cc31dbd/CorsixTH/Lua/entities/patient.lua#L176

(We need a progress of 1 to be 100% sure of the disease)

**Comment by overlordchaos** [comment]
alright. since we have the option to have walk-in alien patients. could it be possible to be developed as a fully dicease? i mean i think its a bad idea to be able to enable a disease that has no way to be diagnosed, you need a high skilled med to send it straight to the cure? its kinda stupid to be able to enable walk-in patients if you are just going to lose them at least most of them.

**Comment by TheCycoONE** [comment]
@overlordchaos yes it's already listed in https://github.com/CorsixTH/CorsixTH/wiki/Programming-Ideas#dna-fixer

Just low priority, unless someone that can write code feels like taking it on.

**Comment by MarkL1961** [comment]
@overlordchaos  
In the campaign it is only in the higher levels and it does take a while before they get diagnosed.
This is why you have training rooms to upskill your doctors.

It can be diagnosed if you have the right staff, I have done it

**Comment by MarkL1961** [comment]
@TheCycoONE   Would it get to one with high skilled staff?

**Comment by overlordchaos** [comment]
@MarkL1961 i have tested it with a hospital with only CONSULTANT doctor's and the result its the same.
we need to think in the original game they where only emergency patients, most likely they run out of time to make it a proper disease. we can fix this and not stick with "ow well in the original this was like that" it was a quickfix cause if you know the history of bullfrog, they where always at the 11th hour rushing the games since they where a small game developer untill EA purchased them and thrashed it apart.

**Comment by TheCycoONE** [comment]
@MarkL1961 yes, lets assume a `MaxDiagDiff` of 0, and a staff that was perfectly skilled and perfectly attentive, and not at all fatigued. If the random number generator was also on our side we would get:

```
diagnosis_base = 0.4
multiplier = 7.5
attn_detail = 1
skill = 1
diag_bonus = 1.4
```
Making the total diagnostic progress 10.9, or 1090% diagnosed just based on the GP. As you can see our 40% from MaxDiagDIff didn't really have much of an impact, even if the difficulty was 1000, we'd be 1050% diagnosed.

The formula feels like it was hacked together - this should impact positively, this should impact negatively, lets add a random factor here and there. If someone were to go to the trouble of comparing I don't think they'd find much correlation with the original game.

**Comment by overlordchaos** [comment]
if anyone can explain to me an easy way to mod the game or casebook so i can assign some exam rooms to the disease (alien DNA) i would be happy to do so and test it on my own.
also i would love to know if it is possible to send the patient to the DNA machine room and after to get threatment from a psychiatrist to get the "cured" part cause for real anyone morphed into an alien would get traumatized. it would be comfortable to know that after the rollback to human the patient was attended by a psychiatrist so he wont commit suicide, making a waste of hospital billing time!

**Comment by TheCycoONE** [comment]
Look at: https://github.com/CorsixTH/CorsixTH/blob/master/CorsixTH/Lua/diseases/alien_dna.lua
For reference on a disease with diagnostic rooms see: https://github.com/CorsixTH/CorsixTH/blob/master/CorsixTH/Lua/diseases/uncommon_cold.lua

The main problem is that there are no graphics for Alien DNA patients using anything. They really shouldn't be coming to the hospital any way except via helicopter, like in the original game - which would be a much easier fix too.

**Comment by overlordchaos** [comment]
@TheCycoONE another solution would be making them step over the DNA machine to give them their humanoid look back, then do some exams then psychiatrist as end goal. in that way you remove the alien issue and you can threat the patient anyways.
as i said my theory is that they run out of time to do the alien desease properly or was a late addition to the game. i suppose it was not meant to be just an emergency sickness considering having a DNA machine with a doctor all the time just in case there is an emergency its stupid. unless they wanted to do a emergency room and they run out of time as well.

**Comment by TheCycoONE** [comment]
Thank you for trying.

**Comment by overlordchaos** [comment]
@TheCycoONE np. happy to help. i think this can be fixed in 2 ways:
1: make the disease only an emergency (as BF did originally) and disable the option to get "walk-in" patients"
2: swap the "alien" appereance (i think the main bug/glitch its with the alien sprite that has missing animations) do a custom one, like the radiation one but in green color. in that way you get all the animations. or just make the alien DNA disease to be the same animation as hard radiation disease.
in my experimentation with other diseases i found out if you add more exam rooms and specially consult's rooms to the disease cure, the game starts to give out errors, and the doctor's/patients get frozen in place. this can be a myor issue if later along the way you plan to add custom diseases. this error can prevent that. unless you can bypass and write from scratch the part of the game that handles the examination/curation of the patient's. that could solve the alien DNA disease. i still theorize the disease was hard code disabled since the disease for some reason never was fully developed.
maybe on the origins of the game it was planned to have an "ER room" with a few diseases only for those rooms, and the DNA alien disease was a leftover from that.

**Comment by TheCycoONE** [comment]
Yes, I agree with option 1. It might land in 0.62.

**Comment by MarkL1961** [comment]
With regards to "They really shouldn't be coming to the hospital any way except via helicopter, like in the original game - which would be a much easier fix too."
This is incorrect, if you played for long enough they did arrive as a patient in later levels, not just emergencies.
Currently in TH they are emergency only patients unless you opt to allow them to arrive like others and when you did opt for this there is a message that pops up telling you about the limitations or at least there was, it may have been removed in the last year or two

**Comment by TheCycoONE** [comment]
@MarkL1961 how did the original deal with the missing graphics then?

**Comment by MarkL1961** [comment]
Most of the time you would not actually notice it as the hospitals tend to get crowded the longer you play.
I was making lots of notes, this was when the project first started so would play on slow speeds a lot of the time

**Comment by TheCycoONE** [comment]
Ew... Ok, thanks MarkL.

**Comment by MarkL1961** [comment]
There are other missing graphics, I think females with broken bones is one where the nurse appears to jump across the room when closing the door of the machine.  
Initially these were being blocked from the game, but then were added as option where the player choses to accept this or not.
There are still issues with the treatment of aliens, I think that is missing the lightening bolt effect at the change back stage in the machine and then the green glow in radiation patients and the explosion when machines blow up.  I don't believe missing animations to be a reason to exclude things though

**Comment by MarkL1961** [comment]
With regards to maxdiag diff, this was something I added many years ago, although it may have been altered since then.
I do remember it was not easy working with something that ranges from 1 - 1000. 1 meaning in level 1 something had a chance of being diagnosed with no other rooms available and at the same time the player getting the choice to send for cure.  Then in later levels it has a much higher count and you have to use diag rooms to get it diagnosed.  It is possible I got something wrong, but it appeared to mimic TH without altering other diseases at the same time.   Hence the use of maths random :)
I also used skill level  (I think it was that) as this was not being used at all in the game, to me it makes sense that a better skilled member of staff would get it right more often than a low skilled junior.

If you can see a better way to mimic TH behaviour feel free to change it

**Comment by mugmuggy** [comment]
I've taken a bit of a look at how TH does some of this logic, around the diagnosis progress.
Patients are initialised with a diagnosis difficulty randomly between 1 and expertise.MaxDiagDiff for their disease
```lua
initialdiagnosisdifficulty = random(1, expertise.MaxDiagDiff)
```
Then for each diagnostic step:
This difficulty value is subtracted away from the calculated diagnosis progress value each step
```lua
diagnosticdiff = currentdiagnosislevel - initialdiagnosisdifficulty 
```
Then there is a random factor calculated as: 
```lua
randomSkill = 250 - randomRange(0,500)
```
Followed by what I'm pretty sure is
```lua
staffSkill = 200
if 'staff' is Junior
  staffSkill = -100
```
Then a tiredness factor
```lua
if staff.fatigue > gbv.VeryTired then
  tired = -200
elseif staff.fatigue > gbv.Tired then
  tired = -100
else
  tired = 100
end
```
I then think if the staff member is *crazy* tired
```lua
quacDoc = -400
```
The room also has an influence, many have no influence, some have positive and some have a negative value, ie x-ray on invis patients. GP doesn't influence it.

The room factors that I can see are with values -400, -200, 0, 200, 400
```lua
roomAdd = roomFactor(disease, room)
```
Add all these values up then add the the staff members ability level, then if that is greater than the current diagnosis level, increase diagnosis progress.
```lua
diagnosticdiff = diagnosticdiff + randomSkill + staffSkill + tired + quacDoc + roomAdd + staff.ability
if diagnosticdiff > currentdiagnosislevel  then
  currentdiagnosislevel = diagnosticdiff
end
```
If it ever reaches 1000, they can be diagnosed.

**Comment by TheCycoONE** [comment]
You are saying the room factor is per disease type then. How did you observe all these components?

**Comment by mugmuggy** [comment]
A lot of reverse engineering :wink: and deducing information from other information, ie I started from looking at the 'hospital' data in the original saves, that lead to some information, then that leads to more information, then you can lean off that to extrapolate to other functionality, sometimes incorrectly.

**Comment by mugmuggy** [comment]
Disease|Scanner|Blood|Cardio|Xray|Ultra|General|Ward|Shrink
---:|---|---|---|---|---|---|---|---
BLOATY_HEAD|400|0|0|0|0|0|200|0
HAIRYITUS|200|0|0|0|0|0|400|0
ELVIS|0|0|0|0|0|0|200|400
INVIS|0|400|0|-200|0|0|0|0
RADIATION|-200|400|0|-400|0|0|0|0
SLACK_TONGUE|400|0|200|0|0|0|0|-200
ALIEN|200|400|0|200|200|0|0|-400
BROKEN_BONES|0|0|200|400|0|0|0|0
BALDNESS|0|400|0|0|0|0|200|-200
DISCRETE_ITCHING|0|400|0|0|0|0|0|0
JELLYITUS|0|0|0|0|0|400|0|0
SLEEPING_ILLNESS|400|0|0|0|0|0|200|0
PREGNANT|0|0|0|0|0|0|0|0
TRANSPARENCY|0|0|0|0|0|0|400|0
UNCOMMON_COLD|0|0|0|0|0|400|0|0
BROKEN_WIND|0|0|0|0|400|0|0|0
SAPRE_RIBS|0|0|0|400|0|0|0|0
KIDNEY_BEANS|0|0|0|200|400|0|0|0
BROKEN_HEART|0|0|400|0|200|0|0|0
RUPTURED_NODULES|400|0|0|0|200|0|0|0
MULTIPLE_TV_PERSONALITIES|0|0|0|0|0|0|0|400
INFECTIOUS_LAUGHTER|0|200|0|0|0|0|0|400
CORRUGATED_ANKLES|0|0|400|0|0|0|0|0
CHRONIC_NOSEHAIR|400|0|0|0|0|0|200|0
3RD_DEGREE_SIDEBURNS|0|0|0|0|0|0|0|400
FAKE_BLOOD|0|400|0|0|0|0|0|0
GASTRIC_EJECTIONS|0|0|0|0|0|0|400|0
THE_SQUITS|0|0|0|0|0|400|200|0
IRON_LUNGS|200|0|0|400|0|0|0|0
SWEATY_PALMS|0|0|-400|0|0|0|0|400
HEAPED_PILES|0|0|0|0|0|0|400|0
GUT_ROT|0|0|0|0|400|0|0|0
GOLF_STONES|0|0|0|200|400|0|0|0
UNEXPECTED_SWELLING|400|0|0|0|0|0|0|0

**Comment by overlordchaos** [comment]
@mugmuggy love all the info you are getting, and love to see you making work TH. ty for your time and skills you use in favor of this project.
that beeing said, could you take a closer look to the alien disease? why it breaks down the game when a patient comes in as a regular patient? and i dont mean the animations, i can live with that. but the doctor's and staff gets frozen in place, rooms get broken or with the patient stuck, etc.

**Comment by mugmuggy** [comment]
I'm working on a big feature, just that this conversation I had just looked at this information the day before and thought I would let everyone know what I had found.

**Comment by MarkL1961** [comment]
@overlordchaos 
I have been trying to replicate the freeze you mention and cannot.  So far out of three aliens the first was diagnosed straight off and the next two sent home - all by consultants.  At the moment I cannot treat the disease yet.
As said previously, you need consultants or highly skilled doctors to diagnose alien dna and even then not every time as there are different strains of most of the diseases in the game.
There is a slider in the hospital policy screen.  If you set this too high patients are more likely to get sent home if the diagnosis process falls below this.  Lower the slider and you are more likely to receive a fax from the doctors.  The ones that say we are 60% sure it is alien dna etc what would you like us to do.  Be warned though as this is a lottery and if you were to send them straight to cure the cure could kill the patient.  Just because it looks like something does not mean it is what you think it is.  This is why the doctors have to do a lot of training :)

**Comment by overlordchaos** [comment]
@MarkL1961 if you desire i can give you a save + my own custom level. all the doctor's get 100% of skills in their field (not sure why i cannot force the game to convert a consultant into psychiatrist+researcher+surgeon at once as you could do on the original game trough training) and also only alien patients (after the emergency comes alone if an emergecy does not come, the aliens wont come alone) and the issue (as you mention) 90% of all the patients, instead of go into diagnosis they are kicked out of the hospital. i think thats an issue it should be adressed and solved. alien DNA its a VISUAL disease. so doctor's can have a minimal idea whats wrong.
ALSO it should be fixed that if an emergency with alien DNA patients does not come first, you dont get no walk-in alien DNA patients. thats a clear bug. (i have tested it on my custom level DISABLING all diseases and only enabled on alien DNA disease on the level) any other disease can start up without an emergency. must be part of hard coded data that prevents the alien dna patients to randomly walk-in without an emergency first.

**Comment by MarkL1961** [comment]
@overlordchaos 
In a new game today 5 out of 5 were diagnosed without any freeze.  This was by a fully skilled consultant.
I had the option to send to research which I did and then the treatment was discovered.  The next alien to arrive was also diagnosed by the same consultant and sent for treatment and still no game freeze.
As previously stated it is expected that you need a very qualified doctor to be able to diagnose this disease.  Otherwise you could staff your hospital with juniors and save a fortune.

Attach your custom level and map, I will take a look

@TheCycoONE in light of the discovery above from @mugmuggy we could add diagnosis room(s), although there would be no animations for this. This would then improve the chances of diagnosis, however having aliens going to a diagnosis room other than GP Office was something I never saw in TH.
The one room that would make sense to me is the blood machine, because apart from getting into the machine when the alien would need to switch to normal the patient is not really visible that much.

**Comment by overlordchaos** [comment]
@MarkL1961 i never said anything about a game freeze. (maybe when i started to tweak stuff but thats not valid cause i might have caused) the issue is that 9 out of 10 patients are kicked out from the hospital trough the GP office upon arrival.

i am thinking the issue here is that i have my custom level with ALL the diseases and machines fully researched from the get-go. maybe if the alien disease its not discovered by research the game wont accept it? that's an issue from anyone that wants to disable research from the get-go of a custom level.
also i think i have missing reasearch items (the enhaced devices on the research lab, on the staff room some items like the arcade game, etc) 
leaving my custom level. if you find a way to tweak the level settings to solve this, you are welcome.
(be warned for the test i DISABLED all the diseases except the alien one, for testing purposes. feel free to enable all of them and tweak the values as you desire them as long the level works once again)
p.s: you might need to trigger (with cheats) an emergency so the first alien patients come in. THEN and only after that, the alien patients will enter trough the hospital if alien disease its the only one enabled. (without the emergency the patients will not come alone)

**Comment by TheCycoONE** [comment]
"that beeing said, could you take a closer look to the alien disease? why it breaks down the game when a patient comes in as a regular patient? and i dont mean the animations, i can live with that. but the doctor's and staff gets frozen in place, rooms get broken or with the patient stuck, etc."

"@MarkL1961 i never said anything about a game freeze."

@overlordchaos these moving goalposts are frustrating. This issue can only be about one thing. Right from the top it looks like it is about diagnosing alien disease, which is exactly what mugmuggy was helping with.

**Comment by overlordchaos** [comment]
@TheCycoONE i now see the misunderstanding. i never meant to call it a "game freeze" aka a game crash.
the Doctor in the GP office, gets stuck in place, and cannot be moved nor deleted as well the room.
thats the "freeze" i was refering about.
and yes its a diagnosis issue. maybe it should not be that difficult to diagnose considering the person is converted into AN ALIEN when they go to the hospital. if you can see the issue you dont need that much diagnosis more than to get the details of the disease.

**Comment by MarkL1961** [comment]
Straight off I already told you to lower maxdiagdiff, you have 900 which means it will still be hard to diagnose.  I have changed to 1

For first walk in a low skilled Dr did not send home and I got this fax

![dna](https://user-images.githubusercontent.com/4965487/35239714-304c8c5a-ffa9-11e7-9b8b-a6e8db29d7a6.png) 

I took a chance and the cure killed him.  I am not a Dr so should not have took the chance.  Now put a Consultant in and he has diagnosed it straight away, but he could not on the next and I got another fax - different strain to the first one - like I said above - it might look like the last patient, but he had a different strain.  I have now altered the hospital policy so that unless the Dr is 80% or more sure he will send patient home - this prevents all the faxes.

Alien DNA is meant to be hard to diagnose, even setting maxdiagdiff to 1 makes it easier but not guaranteed on every patient.  Look at level 1 in the campaign invis is set at 1 and all that does is trigger the faxes where you have to take a chance.

From what I can see the game is working as expected

You should not need the emergency to get aliens.  

You have set no goals so it is impossible to win or lose.

I would not make a hospital that large, I did once and as soon as people tried to play it they said it was difficult to see the edges

**Comment by MarkL1961** [comment]
Your files updated are here https://www.dropbox.com/sh/twba7umxa33m562/AACFkLVcEoSK_EcPYwskykOEa?dl=0

**Comment by overlordchaos** [comment]
@MarkL1961 tyvm for your assistance. i might think its an issue on my game (i have the gog version) and maybe its not corsix but the gog version. will try your fixed files tyvm again for your assistance and time.
also, i did the hospital so expandable, thinking of the future, if corsix team adds more diseases, diagnosis and consultation rooms. not to be used ALL at once. (unless we need 10x more diagnosis and consultation rooms that we have now)
and yes the rules for win or loose are disabled cause my idea for this hospital is to run it as if it was open for decades (thats why its all researched) and its a real hospital. not a game that if you save X patients you win or if you loose X patients you loose the level. i just want to run it even if you save 2k and loose 200 patients. (bad ratio but its an idea)
if i get more issues will try to do a video if you think that will be a berst way to debug issues.

**Comment by MarkL1961** [comment]
You need to have a consultant to diagnose alien dna, the only other dr that can do it is one that is nearly a consultant - look at the skills bar.  The researchers I hired were just above 50% skilled.

That is not really a bug, that is the way it is.

It would be easier if we had diagnosis rooms for alien dna, but that is up for discussion.  If you want to add them for your own use you can edit the lua files, but don't ask for help here if your changes cause a crash/freeze

**Comment by diolemo** [comment]
I'm also experiencing this issue with the default campaign (final mission). 
They always talk to exactly to 1 GP and then immediately go home. 
All my GP are consulants, I have researched the disease and I have the treatment room.

**Comment by overlordchaos** [comment]
> I'm also experiencing this issue with the default campaign (final mission). They always talk to exactly to 1 GP and then immediately go home. All my GP are consulants, I have researched the disease and I have the treatment room.

at this point, just get two point hospital. is on sale with tons of DLC with new and insane diseases. this game is dead this proyect has what 5 years working and most of the GAME bugs that came out in the late 90's are still unsolved. even the original devs give out an interview when they helped out making two point hospital, and they recognized that there was tons of bugs, that the easy way to fix it was cut short levels and diseases. so if they could not fix it i doubt in 2025 will get fixed.

**Comment by ARGAMX** [comment]
> 
> at this point, just get two point hospital. is on sale with tons of DLC with new and insane diseases. this game is dead this proyect has what 5 years working and most of the GAME bugs that came out in the late 90's are still unsolved. even the original devs give out an interview when they helped out making two point hospital, and they recognized that there was tons of bugs, that the easy way to fix it was cut short levels and diseases. so if they could not fix it i doubt in 2025 will get fixed.

Can you please provide a specific list of things that haven't been unsolved since the original game released in 90x? Fixing which would stop this project from being called dead?

**Comment by lewri** [comment]
> this game is dead this proyect has what 5 years working

I'm not sure what you mean in reference to this project struggling at all. CorsixTH continues to make progress in recreating the game in a new engine and there is no signals that would stop.

However these recent comments are not helpful to the topic nor the project, so I will lock further comments in this issue and look into @diolemo 's report separately.
